spec/hcloud/firewall_spec.rb

```diff
@@ -1,212 +1,102 @@
-  end
-
-  it '#apply_to_resources, missing apply_to' do
-    expect { client.firewalls['fw'].apply_to_resources(apply_to: nil) }.to(
-      raise_error(Hcloud::Error::InvalidInput)
-    )
-  end
-
-  it '#remove_from_resources' do
-    remove_from = [{
-      'server' => {
-        'id' => 1
-      },
-      'type' => 'server'
-    }]
-    actions = client.firewalls['fw'].remove_from_resources(remove_from: remove_from)
-
-    expect(client.firewalls['fw'].applied_to.length).to eq(1)
-    expect(client.firewalls['fw'].applied_to[0][:server][:id]).to eq(42)
-
-    expect(client.actions.count).to eq(3)
-    expect(client.firewalls['fw'].actions.count).to eq(3)
-    expect(actions[0].command).to eq('remove_from_resources')
-  end
-
-  it '#remove_from_resources, missing remove_from' do
-    expect { client.firewalls['fw'].remove_from_resources(remove_from: nil) }.to(
-      raise_error(Hcloud::Error::InvalidInput)
-    )
-  end
+describe Hcloud::Firewall, doubles: :firewall do
+  include_context 'test doubles'
 
-  it '#update(name:)' do
-    id = client.firewalls['fw'].id
-    expect(id).to be_a Integer
-    expect(client.firewalls.find(id).name).to eq('fw')
-    expect(client.firewalls.find(id).update(name: 'firewall').name).to eq('firewall')
-    expect(client.firewalls.find(id).name).to eq('firewall')
+  let :firewalls do
+    Array.new(Faker::Number.within(range: 20..150)).map { new_firewall }
   end
 
-  it '#update(labels:)' do
-    id = client.firewalls['firewall'].id
-    firewall = client.firewalls[id]
-    updated = firewall.update(labels: { 'source' => 'update' })
-    expect(updated.labels).to eq({ 'source' => 'update' })
-    expect(client.firewalls[id].labels).to eq({ 'source' => 'update' })
-  end
+  let(:firewall) { firewalls.sample }
 
-  it '#where -> find by label_selector' do
-    firewalls = client.firewalls.where(label_selector: 'source=update').to_a
-    expect(firewalls.length).to eq(1)
-    expect(firewalls.first.labels).to include('source' => 'update')
+  let :client do
+    Hcloud::Client.new(token: 'secure')
   end
 
-  it '#destroy' do
-    expect(client.firewalls.first).to be_a Hcloud::Firewall
-    id = client.firewalls.first.id
-    expect(id).to be_a Integer
-    expect(client.firewalls.find(id).destroy).to be_a Hcloud::Firewall
-    expect(client.firewalls[id]).to be nil
+  include_examples 'it_supports_fetch', described_class
+  include_examples 'it_supports_find_by_id_and_name', described_class
+  include_examples 'it_supports_update', described_class, { name: 'new_name' }
+  include_examples 'it_supports_destroy', described_class
+  include_examples 'it_supports_labels_on_update', described_class
+
+  context '#create' do
+    it 'handle missing name' do
+      expect { client.firewalls.create(name: nil) }.to(
+        raise_error(Hcloud::Error::InvalidInput)
+      )
+    end
+
+    context 'works' do
+      it 'with only name' do
+        params = { name: 'moo', labels: { 'key' => 'value' } }
```

Comment:
the test-name is a bit misleading if I look at the params, because you provide name AND labels. Is this intended?